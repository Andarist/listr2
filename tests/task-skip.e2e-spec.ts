import { Listr } from '@root/index'

describe('skip a task', () => {
  let log: jest.SpyInstance<void, string[][]>
  let info: jest.SpyInstance<void, string[][]>

  beforeEach(async () => {
    log = jest.spyOn(console, 'log').mockImplementation()
    info = jest.spyOn(console, 'info').mockImplementation()
  })

  afterEach(async () => {
    jest.clearAllMocks()
  })

  it('should skip the task from internal call', async () => {
    await new Listr(
      [
        {
          task: async (ctx, task): Promise<void> => {
            task.skip('skipped')
          }
        }
      ],
      { renderer: 'verbose' }
    ).run()

    expect(log).toBeCalledWith('[STARTED] Task without title.')
    expect(info).toBeCalledWith('[SKIPPED] skipped')
  })

  it('should skip the task from skip method', async () => {
    await new Listr(
      [
        {
          skip (): string {
            return 'skipped'
          },
          // eslint-disable-next-line @typescript-eslint/no-empty-function
          task: async (): Promise<void> => {}
        }
      ],
      { renderer: 'verbose' }
    ).run()

    expect(log).toBeCalledWith('[STARTED] Task without title.')
    expect(info).toBeCalledWith('[SKIPPED] skipped')
  })

  it('should skip the task from async skip method returning either boolean or string', async () => {
    await new Listr(
      [
        {
          skip: async(): Promise<boolean | string> => {
            await new Promise(r => setTimeout(r, 50));
            
            if (Math.random() < 0.5) {
              return true;
            }
            
            return 'skipped';
          },
          // eslint-disable-next-line @typescript-eslint/no-empty-function
          task: async (): Promise<void> => {}
        }
      ],
      { renderer: 'verbose' }
    ).run()

    expect(log).toBeCalledWith('[STARTED] Task without title.')
    expect(info).toBeCalledWith('[SKIPPED] Skipped task without a title.')
  })
  
  it('should skip the task from async skip method returning boolean', async () => {
    await new Listr(
      [
        {
          skip: async(): Promise<boolean> => {
            await new Promise(r => setTimeout(r, 50));
            
            return true;
          },
          // eslint-disable-next-line @typescript-eslint/no-empty-function
          task: async (): Promise<void> => {}
        }
      ],
      { renderer: 'verbose' }
    ).run()

    expect(log).toBeCalledWith('[STARTED] Task without title.')
    expect(info).toBeCalledWith('[SKIPPED] Skipped task without a title.')
  })
  
  it('should skip the task from async skip method returning string', async () => {
    await new Listr(
      [
        {
          skip: async(): Promise<string> => {
            await new Promise(r => setTimeout(r, 50));
            
            return 'skipped'
          },
          // eslint-disable-next-line @typescript-eslint/no-empty-function
          task: async (): Promise<void> => {}
        }
      ],
      { renderer: 'verbose' }
    ).run()

    expect(log).toBeCalledWith('[STARTED] Task without title.')
    expect(info).toBeCalledWith('[SKIPPED] skipped')
  })
  
  it('skip to enable by context will work properly in serial', async () => {
    await new Listr(
      [
        {
          task: async (ctx, task): Promise<void> => {
            ctx.test = true
            task.skip('skipped')
          }
        },
        {
          enabled: (ctx): boolean => ctx.test,
          task: async (ctx, task): Promise<void> => {
            task.output = 'enabled'
          }
        }
      ],
      { renderer: 'verbose', concurrent: false }
    ).run()

    expect(log).toBeCalledWith('[STARTED] Task without title.')
    expect(info).toBeCalledWith('[SKIPPED] skipped')
    expect(info).toBeCalledWith('[DATA] enabled')
  })

  it('skip to enable by context will not work properly in concurrent', async () => {
    await new Listr(
      [
        {
          task: async (ctx, task): Promise<void> => {
            ctx.test = true
            task.skip('skipped')
          }
        },
        {
          enabled: (ctx): boolean => ctx.test,
          task: async (ctx, task): Promise<void> => {
            task.output = 'enabled'
          }
        }
      ],
      { renderer: 'verbose', concurrent: true }
    ).run()

    expect(log).toBeCalledWith('[STARTED] Task without title.')
    expect(info).toBeCalledWith('[SKIPPED] skipped')
    expect(info).not.toBeCalledWith('[DATA] enabled')
  })
})
