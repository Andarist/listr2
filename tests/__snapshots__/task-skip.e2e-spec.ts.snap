// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`skip a task should skip the task from async skip method returning either boolean or string: dKEmraxPfJH3cEj4bHeTsieEALbqcFIR-exit 1`] = `[]`;

exports[`skip a task should skip the task from async skip method returning either boolean or string: dKEmraxPfJH3cEj4bHeTsieEALbqcFIR-stderr 1`] = `[]`;

exports[`skip a task should skip the task from async skip method returning either boolean or string: dKEmraxPfJH3cEj4bHeTsieEALbqcFIR-stdout 1`] = `
[
  "{"event":"STATE","data":"STARTED","task":{"hasFinalized":false}}
",
  "{"event":"MESSAGE","data":{"skip":"Skipped task without a title."},"task":{"hasFinalized":false}}
",
  "{"event":"STATE","data":"SKIPPED","task":{"hasFinalized":true}}
",
]
`;

exports[`skip a task should skip the task from internal call: xbUV7hvMZiVL4dEYDyiXKFhpTNo9ASiA-exit 1`] = `[]`;

exports[`skip a task should skip the task from internal call: xbUV7hvMZiVL4dEYDyiXKFhpTNo9ASiA-stderr 1`] = `[]`;

exports[`skip a task should skip the task from internal call: xbUV7hvMZiVL4dEYDyiXKFhpTNo9ASiA-stdout 1`] = `
[
  "{"event":"STATE","data":"STARTED","task":{"hasFinalized":false}}
",
  "{"event":"STATE","data":"SKIPPED","task":{"hasFinalized":true}}
",
  "{"event":"MESSAGE","data":{"skip":"skipped"},"task":{"hasFinalized":true}}
",
]
`;

exports[`skip a task should skip the task from skip method: uOy7d3bwwmW4LXWPY3HDh4k25LDwa5RZ-exit 1`] = `[]`;

exports[`skip a task should skip the task from skip method: uOy7d3bwwmW4LXWPY3HDh4k25LDwa5RZ-stderr 1`] = `[]`;

exports[`skip a task should skip the task from skip method: uOy7d3bwwmW4LXWPY3HDh4k25LDwa5RZ-stdout 1`] = `
[
  "{"event":"STATE","data":"STARTED","task":{"hasFinalized":false}}
",
  "{"event":"MESSAGE","data":{"skip":"skipped"},"task":{"hasFinalized":false}}
",
  "{"event":"STATE","data":"SKIPPED","task":{"hasFinalized":true}}
",
]
`;

exports[`skip a task skip to enable by context will not work properly in concurrent: WKhUc2IR0U5Hc7hKFuvFUq36ovxjyPy6-exit 1`] = `[]`;

exports[`skip a task skip to enable by context will not work properly in concurrent: WKhUc2IR0U5Hc7hKFuvFUq36ovxjyPy6-stderr 1`] = `[]`;

exports[`skip a task skip to enable by context will not work properly in concurrent: WKhUc2IR0U5Hc7hKFuvFUq36ovxjyPy6-stdout 1`] = `
[
  "{"event":"STATE","data":"STARTED","task":{"hasFinalized":false}}
",
  "{"event":"STATE","data":"SKIPPED","task":{"hasFinalized":true}}
",
  "{"event":"MESSAGE","data":{"skip":"skipped"},"task":{"hasFinalized":true}}
",
]
`;

exports[`skip a task skip to enable by context will work properly in serial: CR5tEdrMVapfqwMjIa4nP1CgEYZu3tfc-exit 1`] = `[]`;

exports[`skip a task skip to enable by context will work properly in serial: CR5tEdrMVapfqwMjIa4nP1CgEYZu3tfc-stderr 1`] = `[]`;

exports[`skip a task skip to enable by context will work properly in serial: CR5tEdrMVapfqwMjIa4nP1CgEYZu3tfc-stdout 1`] = `
[
  "{"event":"STATE","data":"STARTED","task":{"hasFinalized":false}}
",
  "{"event":"STATE","data":"SKIPPED","task":{"hasFinalized":true}}
",
  "{"event":"MESSAGE","data":{"skip":"skipped"},"task":{"hasFinalized":true}}
",
  "{"event":"STATE","data":"STARTED","task":{"hasFinalized":false}}
",
  "{"event":"OUTPUT","data":"enabled","task":{"hasFinalized":false}}
",
  "{"event":"STATE","data":"COMPLETED","task":{"hasFinalized":true}}
",
]
`;
